diff --git a/BUILD.gn b/BUILD.gn
index 277f47717d..1d301531fd 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -2685,6 +2685,7 @@ v8_header_set("v8_headers") {
     "include/v8-wasm.h",
     "include/v8-weak-callback-info.h",
     "include/v8.h",
+    "include/wasm-v8-lowlevel.hh",
   ]
 
   sources += [
@@ -3700,6 +3702,7 @@ v8_header_set("v8_internal_headers") {
 
   if (v8_enable_webassembly) {
     sources += [
+      "src/wasm-v8-lowlevel.cc",
       "src/asmjs/asm-js.h",
       "src/asmjs/asm-names.h",
       "src/asmjs/asm-parser.h",
